### PR TITLE
AP_JSON: added JSON parsing library

### PR DESCRIPTION
**Pull request by tridge**
This is an alternative to #23673 that adds AP_JSON based on a re-working of picojson. It runs on both SITL and HAL_ChibiOS

I think this provides a nicer JSON API which is better for future usage of JSON

### CODE REVIEW

**Comment by rmackay9** [comment]
This could be good for integration with various AI systems that all seem to use JSON..

**Comment by tridge** [comment]
> This could be good for integration with various AI systems that all seem to use JSON..

I'd expect those to use a json implementation in lua (JSON is very easy in lua)

File: Tools/ardupilotwaf/chibios.py [hunk 1]
```diff
@@ -143,6 +144,26 @@ def run(self):
...
+
+    if not task.env.vehicle_binary or task.env.SIM_ENABLED:
+        # we only want to check symbols for vehicle binaries, allowing examples
+        # to use C++ exceptions. We also allow them in simulator builds
+        return
+
+    # we use string find on these symbols, so this catches all types of throw calls
+    blacklist = ['std::__throw']
+
+    print("COMMAND: %s -C %s" % (task.env.get_flat('NM'), elfpath))
```
Comment (tridge) [inline Tools/ardupilotwaf/chibios.py:161]: this print() should be removed

**Review by davidbuzz — APPROVED** [review]
looks good to me once you explained the gmtime->gmtime_r changes

**Review by peterbarker — APPROVED** [review]
Good so long as the Callisto model works on SITL-on-hardware.

File: Tools/ardupilotwaf/chibios.py [hunk 2]
```diff
@@ -143,6 +144,26 @@ def run(self):
...
+    check for disallowed symbols in elf file, such as C++ exceptions
+    '''
+    elfpath = task.inputs[0].abspath()
+
+    if not task.env.vehicle_binary or task.env.SIM_ENABLED:
+        # we only want to check symbols for vehicle binaries, allowing examples
+        # to use C++ exceptions. We also allow them in simulator builds
+        return
+
+    # we use string find on these symbols, so this catches all types of throw calls
```
Comment (tridge) [inline Tools/ardupilotwaf/chibios.py:158]: 'we believe' - could it be inlined?

**Comment by Ryanf55** [comment]
Take a look at the benefit of `std::array`: https://godbolt.org/z/GraMsY3qP

At compile time, you can get an error when trying to access values out of range. Peter proposes doing a trivial example of changing a raw array to `std::array` and examining the binaries. 

I would recommend enabling this in SITL when SITL is in debug mode. Coupled with using `std::array` for any arrays with size known at compile time, it allows for huge improvements in compile-time-safety with regards to array access.

**Comment by tridge** [comment]
tested on CubeOrange with Callisto model:
image

File: Tools/ardupilotwaf/chibios.py (hunk 2 above)
Comment (tridge) [reply Tools/ardupilotwaf/chibios.py]: done

**Pull request merged**